Add join and leave messages

tox-user opened the pull request:
Adds messages that notify when user joins or leaves group chat or changes their name. Implements #4128.

Current issues:
- ~~when user joins a non empty group chat they will always get a join message for every participant at once~~
- ~~every time someone joins a group that we are in, we don't see their user name at first. Instead we see "Tox User" as their name. A few seconds later the name changes to their real name, which creates additional message~~

~~I am not sure how to fix them yet.~~

Depends on https://github.com/TokTok/c-toxcore/pull/819

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/4923)
<!-- Reviewable:end -->

iphydf commented:
Hi tox-user, we're working on a change that slightly changes the group chat api. Can we discuss somewhere? </hijack-thread>

tox-user commented:
Hi @iphydf could you add me on Tox? You will find my ID in my profile.

tox-user committed:
feat(groupchat): add join and leave messages

Implements #4128
Adds messages that notify when user joins or leaves group chat or changes their name.

tox-user commented:
This is ready for testing and review now.

Diadlo commented:
1. It doesn't works
2. If you move all code in `if (oldPeersList.length() > 0)` (with `if` itself) after peer filling but before `emit` it will work
3. And after #4978 merge (because otherwise you haven't any messages on first peer come)
4. And since we show friend alias if we have peer as friend, probably you shouldn't `emit peerNameChanged` if friend found (probably, we should show both and correct handle it, but feel free to leave it now)

tox-user commented:
It's true that it doesn't work in this scenario:
1. Client `#1` creates a group and invites client `#2`.
2. Client `#2` joins.
3. Client `#1` will not receive the join notification.

PR https://github.com/qTox/qTox/pull/4978 itself doesn't fix it. It makes it more strange, because then client `#2` gets duplicate join notifications.

Diadlo commented:
@tox-user Did you move `if (oldPeersList.length() > 0)`?

tox-user commented:
That helped, but I get a notification spam when I join a bigger existing group.

Talkless commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/model/group.cpp, line 53 at r1](https://reviewable.io/reviews/qtox/qtox/4923#-LDRF4HH7Ew9iecz63Ap:-LDRF4HI9i_vnRX0itPw:bc5g0lq) ([raw file](https://github.com/qtox/qtox/blob/ee9c68e29b0ac7dc4046f3346e476b5e47431d8d/src/model/group.cpp#L53)):*
> ```cpp
>     // which is later changed to their real name. Because of that we treat everyone
>     // who changed the name from empty to non empty as a newly joined user
>     if (name != peers[peerId]) {
> ```

Same `peers[peerId]` lookup is performed two-three times, could it be avoided?

---

*[src/model/group.cpp, line 109 at r1](https://reviewable.io/reviews/qtox/qtox/4923#-LDRFZ0ZCxLXBsd8Kcbl:-LDRFZ0ZCxLXBsd8Kcbm:bbqvuk2) ([raw file](https://github.com/qtox/qtox/blob/ee9c68e29b0ac7dc4046f3346e476b5e47431d8d/src/model/group.cpp#L109)):*
> ```cpp
>     if (oldPeersList.length() > 0) {
>         // peers that have joined
>         for (int i = 0; i < peers.length(); i++) {
> ```

is `int i` index really needed here? As far as I can see, ranged-based for loop could work?

---

*[src/model/group.cpp, line 118 at r1](https://reviewable.io/reviews/qtox/qtox/4923#-LDRFhIj7046JP_BqrMR:-LDRFhIk2eT6-MwAPBlv:bfnu268) ([raw file](https://github.com/qtox/qtox/blob/ee9c68e29b0ac7dc4046f3346e476b5e47431d8d/src/model/group.cpp#L118)):*
> ```cpp
> 
>         // peers that have left
>         for (int i = 0; i < oldPeersList.length(); i++) {
> ```

range-based for loop instead?

---

*[src/widget/widget.cpp, line 1841 at r1](https://reviewable.io/reviews/qtox/qtox/4923#-LDRFnI91RuK1n7GAjNV:-LDRFnI91RuK1n7GAjNW:b-4aputg) ([raw file](https://github.com/qtox/qtox/blob/ee9c68e29b0ac7dc4046f3346e476b5e47431d8d/src/widget/widget.cpp#L1841)):*
> ```cpp
>     const QString message = tr("%1 has joined the chat");
> 
>     Group* g = GroupList::findGroup(groupnumber);
> ```

`Group* const g`? Or maybe even `const Group* const g` (if getId() is const)?

---

*[src/widget/widget.cpp, line 1855 at r1](https://reviewable.io/reviews/qtox/qtox/4923#-LDRG0qhD9kFmmgej0wX:-LDRG0qhD9kFmmgej0wY:b-hfk38u) ([raw file](https://github.com/qtox/qtox/blob/ee9c68e29b0ac7dc4046f3346e476b5e47431d8d/src/widget/widget.cpp#L1855)):*
> ```cpp
>     const QString message = tr("%1 has left the chat");
> 
>     Group* g = GroupList::findGroup(groupnumber);
> ```

same constsness.

---

*[src/widget/widget.cpp, line 1869 at r1](https://reviewable.io/reviews/qtox/qtox/4923#-LDRG3-72OdOqUsuMh1e:-LDRG3-72OdOqUsuMh1f:b3dnjut) ([raw file](https://github.com/qtox/qtox/blob/ee9c68e29b0ac7dc4046f3346e476b5e47431d8d/src/widget/widget.cpp#L1869)):*
> ```cpp
>     const QString message = tr("%1 is now known as %2");
> 
>     Group* g = GroupList::findGroup(groupnumber);
> ```

same constness

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/4923)*
<!-- Sent from Reviewable.io -->

tox-user commented:
@Talkless thanks for reviewing, but unfortunately this PR doesn't work anymore, because of https://github.com/qTox/qTox/pull/4978. I don't know how to get it working again, so I've decided to leave it and move on to other things. Maybe I will take a look at it again some time in the future or someone else will pick it up.
